api/src/pcapi/core/external/compliance.py

```diff
@@ -0,0 +1,130 @@
+
+
+def update_offer_compliance_score(offer: offers_models.Offer, is_primary: bool, asynchronous: bool = True) -> None:
+    payload = _get_payload_for_compliance_api(offer)
+    if is_primary:
+        if asynchronous:
+            compliance_tasks.update_offer_compliance_score_primary_task.delay(payload)
+        else:
+            compliance_tasks.update_offer_compliance_score_primary_task(payload)
+    else:
+        if asynchronous:
+            compliance_tasks.update_offer_compliance_score_secondary_task.delay(payload)
+        else:
+            compliance_tasks.update_offer_compliance_score_secondary_task(payload)
+
+
+def make_update_offer_compliance_score(payload: GetComplianceScoreRequest) -> None:
+    offer = offers_models.Offer.query.get(payload.offer_id)
+
+    client_id = settings.COMPLIANCE_API_CLIENT_ID
+
+    if settings.IS_DEV:
+        # This is a fake auth_token; if you want to test it in development env
+        # you have to use the result returned by 'gcloud auth print-access-token'
+        auth_token = "gcloud auth print-access-token"
+
+        try:
+            id_token_response = requests.post(
+                f"https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/{settings.COMPLIANCE_API_SERVICE_ACCOUNT}:generateIdToken",
+                headers={
+                    "Authorization": f"Bearer {auth_token}",
+                    "Content-Type": "application/json; charset=utf-8",
+                },
+                json={"audience": client_id, "includeEmail": "true"},
+            )
+            id_token = id_token_response.json()["token"]
+        except requests.exceptions.InvalidHeader:
+            # The auth_token has probably not been changed to a correct value
+            logger.info("Check the value of your auth_token")
+            data_score = None
+            return
+        except Exception as exc:  # pylint: disable=broad-except
+            logger.info("Could not get id_token", extra={"exc": exc})
+            data_score = None
+            return
+
+    else:
+        id_token = auth_api.get_id_token_from_google(client_id)
+
+    data = payload.to_dict()
+    try:
+        api_response = requests.post(
+            "https://compliance.passculture.team/latest/model/compliance/scoring",
+            headers={
+                "Authorization": f"Bearer {id_token}",
+                "accept": "application/json",
+                "Content-Type": "application/json",
+            },
+            json=data,
+        )
+        data_score = api_response.json()["probability_validated"]
+
+    except Exception as exc:  # pylint: disable=broad-except
+        logger.exception("Connection to API Compliance failed", extra={"exc": exc})
+        data_score = None
+
+    if data_score:
+        if offer.extraData is None:
+            offer.extraData = {}
+        offer.extraData["complianceScore"] = data_score
+    db.session.add(offer)
+    db.session.commit()
```

Comment:
Si on ne passe pas dans le `if` au-dessus, que va-t-on `commit` en base de données ?

Reply:
En fait j'ai pas besoin de faire ce commit ici, ni le add, si je ne m'abuse. Ils sont déjà faits plus tard dans la création d'offre.

Reply:
Lorsque la fonction est appelée de façon asynchrone, alors le `commit` dans la création d'offre est déjà fait (c'est plus tard par rapport à la création de la _cloud task_, mais plus plus tôt que son appel). Si tu n'en mets aucun ici, alors rien ne sera enregistré dans la _cloud task_.

Je me demande même s'il ne faudrait pas faire un select for update afin de s'assurer que l'enregistrement dans la route et la _cloud task_ ne se percutent pas, un peu comme vient de le faire @damien-ramelet pour une autre situation de concurrence.

Par contre, en effet comme tu permets de l'appeler de façon synchrone, n'y a-t-il pas un risque de `commit` anticipé ? Faudrait-il conditionner le commit en fonction des paramètres ?